### PR TITLE
Render layers as groups

### PR DESCRIPTION
**Pull request by nyalldawson**
This exposes the ability to render group layers, as described by qgis/QGIS-Enhancement-Proposals#235.

When a group is selected in the layer tree the styling dock will show a new panel for controlling the appearance of the group as a whole, with options for group opacity, blend mode and layer effects.

image

Sponsored by Andrew Fletcher.

### CODE REVIEW

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -12241,9 +12243,15 @@ void QgisApp::removeLayer()
...
 
-  const auto constSelectedNodes = selectedNodes;
-  for ( QgsLayerTreeNode *node : constSelectedNodes )
+  for ( QgsLayerTreeNode *node : selectedNodes )
   {
+    if ( QgsLayerTreeGroup *group = qobject_cast< QgsLayerTreeGroup * >( node ) )
+    {
+      if ( QgsGroupLayer *groupLayer = group->groupLayer() )
+      {
+        QgsProject::instance()->removeMapLayer( groupLayer );
```
Comment (3nids) [inline src/app/qgisapp.cpp:12252]: could this be an issue regarding the order of signal here ?
`layerWillBeRemoved` and `layerRemoved` will both be sent first for the layer group, and later for each layer.

Comment (nyalldawson) [reply src/app/qgisapp.cpp:12252]: I think that should be fine -- the group layers don't have any ownership of the other layers, so to me it makes sense that they get cleaned up first

Comment (3nids) [reply src/app/qgisapp.cpp:12252]: ok, thanks for the answer

File: src/core/project/qgsprojectutils.cpp [hunk 2]
```diff
@@ -52,3 +53,18 @@ bool QgsProjectUtils::updateLayerPath( QgsProject *project, const QString &oldPa
   }
   return res;
 }
+
+bool QgsProjectUtils::layerIsContainedInGroupLayer( QgsProject *project, QgsMapLayer *layer )
+{
+  const QMap<QString, QgsMapLayer *> mapLayers( project->mapLayers() );
```
Comment (nirvn) [inline src/core/project/qgsprojectutils.cpp:59]: @nyalldawson , just an idea (not part of this PR), it might be useful to eventually add a QgsProject::mapLayersByType( Qgis::QgsMapLayerType ) function. It'd be useful here (and likely in other places too).

**Review by nirvn — APPROVED** [review]

File: src/core/project/qgsprojectutils.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/project/qgsprojectutils.cpp]: good catch -- there's already a method for that. I've pushed a commit to use it

**Comment by agiudiceandrea** [comment]
See https://github.com/qgis/QGIS/issues/53057, https://github.com/qgis/QGIS/issues/53088, https://github.com/qgis/QGIS/issues/51337.

**Comment by Gustry** [comment]
And #51393

**Comment by agiudiceandrea** [comment]
Please have a look at https://github.com/qgis/QGIS/issues/57249.
